### PR TITLE
Release artifacts validation

### LINKED ISSUE #562: Provide a script for installing the CLI from the archive files we publish as a part of releases
**Describe the feature request**

As a user, I should be able to run a script that performs the multi-step process of installing the CLI:

- download and untar the appropriate tanzu-framework-(OS)-(ARCH).tar
- `cd` to the tar destination directory:
- move/copy core/v0.1.0/tanzu_core-(OS)-(ARCH) into somewhere in your PATH and rename it `tanzu`
- `export TANZU_CLI_NO_INIT=true`
- `tanzu plugin install --local ./ all`
- `tanzu plugin list` show the following plugins installed: "login", "pinniped-auth", "cluster", "management-cluster", "kubernetes-release", "package"

Community Edition includes scripts that accomplish this:

- https://github.com/vmware-tanzu/community-edition/blob/main/hack/install.bat
- https://github.com/vmware-tanzu/community-edition/blob/main/hack/install.sh

**Describe alternatives you've considered**

We should also find a better way to distribute the core CLI binary (#3) and provide an API for CLI plugin discovery (#74). However, this is intended as a stop-gap until that's available.

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [x] CLI
- [x] Docs
- [x] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

### LINKED ISSUE #1134: Validate release artifacts before publishing
During a release we test the release candidate commit, tag it on main branch and publish a release. In this process we do not test the release artifacts which contain signed MacOS and Windows binaries.

We need automated tests to validate release artifacts before publishing them. These can be wired in a Github Actions workflow which can be triggered through `gh` CLI before publishing the release similar to how it is done in https://github.com/vmware-tanzu/community-edition/pull/2304

### PR DESCRIPTION
**Pull request by rajaskakodkar**
### What this PR does / why we need it
This adds automation to validate release artifacts which can be triggered before publishing the release

### Which issue(s) this PR fixes

Fixes #1134

### Describe testing done for PR
These workflows have been tested here:
https://github.com/rajaskakodkar/tanzu-framework/actions/runs/1451790118 -> MacOS and Linux
https://github.com/rajaskakodkar/tanzu-framework/actions/runs/1451789814 -> Windows OS

EDIT:
Tested these workflows with release v0.11.1
https://github.com/rajaskakodkar/tanzu-framework/actions/runs/1498809428 -> MacOS and Linux
https://github.com/rajaskakodkar/tanzu-framework/actions/runs/1498808654 -> Windows OS

EDIT:
Tested these workflows with release v0.12.0
Windows: https://github.com/rajaskakodkar/tanzu-framework/runs/4350652103?check_suite_focus=true
MacOS and Linux: https://github.com/rajaskakodkar/tanzu-framework/runs/4350650498?check_suite_focus=true

### Release note

```release-note
Added validation checks for release artifacts
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### Additional information
These workflows can be triggered by the following commands 
```
gh workflow run --repo https://github.com/vmware-tanzu/tanzu-framework/ --ref release-artifacts-validation test-release-artifacts-windows.yaml -f release_version=v0.10.0

gh workflow run --repo https://github.com/vmware-tanzu/tanzu-framework/ --ref release-artifacts-validation test-release-artifacts-linux-macos.yaml -f release_version=v0.10.0
```
#### Special notes for your reviewer
We need to merge #1157 before we merge this.

### CODE REVIEW

**Comment by anujc25** [comment]
We might need to change some of this based on #1144 PR.
I am working on updating the documentation and installation guide as part of PR #1188

**Comment by rajaskakodkar** [comment]
Update:
I have updated this PR with the changes required after API driven CLIPlugin Discovery and tested the same with release v0.11.1

Logs:
https://github.com/rajaskakodkar/tanzu-framework/actions/runs/1498809428 -> MacOS and Linux
https://github.com/rajaskakodkar/tanzu-framework/actions/runs/1498808654 -> Windows OS

I think we are good to get this in after #1157

File: hack/test-release-artifacts.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+if [ "${OS}" == 'darwin' ]; then
+    for binary in "${temp_dir}"/tanzu/cli/*; do
+        if [[ -d "${binary}" ]]; then
+            spctl -vv --type install --asses "${binary}"/"${version}"/*
+        fi
+    done
+fi
+
+./hack/install.sh "${temp_dir}"/tanzu "${version}"
```
Comment (navidshaikh) [inline hack/test-release-artifacts.sh:38]: per comment on https://github.com/vmware-tanzu/tanzu-framework/pull/1157#discussion_r758000121

this becomes
```
./hack/install.sh "${temp_dir}/tanzu/cli/core/${version}/tanzu-core-${OS}_${ARCH}"
```

**Review by navidshaikh — APPROVED** [review]
lgtm! left a related review comment from #1157

**Comment by rajaskakodkar** [comment]
Tested the updated scripts on my fork and here are the results:
Windows: https://github.com/rajaskakodkar/tanzu-framework/runs/4350652103?check_suite_focus=true
MacOS and Linux: https://github.com/rajaskakodkar/tanzu-framework/runs/4350650498?check_suite_focus=true

**Comment by rajaskakodkar** [comment]
Noticed this [warning](https://github.com/rajaskakodkar/tanzu-framework/runs/4350652103?check_suite_focus=true#step:3:3320)
```
Warning: Failed to initialize plugin '"management-cluster"' after installation.
```
and [error](https://github.com/rajaskakodkar/tanzu-framework/runs/4350652103?check_suite_focus=true#step:3:3321)
```
Error: unable to ensure prerequisites: unable to ensure tkg BOM file: failed to download TKG compatibility file from the registry: failed to list TKG compatibility image tags: Get "https://projects-stg.registry.vmware.com/v2/": x509: certificate signed by unknown authority
```
on windows and not on macOS or linux.

However, the installation of cli and plugins was successful on windows

cc: @anujc25

**Review by navidshaikh — APPROVED** [review]
lgtm!

**Comment by navidshaikh** [comment]
> Warning: Failed to initialize plugin '"management-cluster"' after installation.

@rajaskakodkar seems like a temp BoM image download flake ? should go away with a retry.

**Comment by rajaskakodkar** [comment]
@navidshaikh this has been persistent on multiple retries so doesn't look like a flake to me. Also, I have noticed it only on Windows.

**Review by vuil — APPROVED** [review]
lgtm, thanks!

**Pull request merged**